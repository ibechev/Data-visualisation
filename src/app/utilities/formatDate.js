export const formatDate = date => {
  const months = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec"
  ];
  const tempDate = date.split("-");
  return `${months[parseInt(tempDate[1]) - 1]} ${tempDate[0].substring(2)}`;
};
