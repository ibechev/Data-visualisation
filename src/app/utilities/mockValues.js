export const mockAlerts = [
  {
    name: "Adam Campton",
    event: "end of year review",
    eventDate: "2019-01-18",
    importance: "very"
  },
  {
    name: "Emma O'Sullivan",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Brandon Walsh",
    event: "end of year review",
    eventDate: "2019-02-06",
    importance: "middle"
  },
  {
    name: "David O'Brien",
    event: "end of year review",
    eventDate: "2019-02-23",
    importance: "very"
  },
  {
    name: "Sophie O'Connor",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Sean Connor",
    event: "end of year review",
    eventDate: "2019-01-05",
    importance: "very"
  },
  {
    name: "Gery Brown",
    event: "end of year review",
    eventDate: "2019-01-12",
    importance: "middle"
  },
  {
    name: "Stephen Greenday",
    event: "end of year review",
    eventDate: "2019-02-11",
    importance: "very"
  },
  {
    name: "Lisa Aniston",
    event: "end of year review",
    eventDate: "2018-12-27",
    importance: "low"
  },
  {
    name: "Gordon Niels",
    event: "end of year review",
    eventDate: "2018-12-28",
    importance: "middle"
  }
];

export const mockCustomerInfo = {
  max: 270,
  data: [
    { month: "Dec-2017", value: 99 },
    { month: "Jan-2018", value: 80 },
    { month: "Feb-2018", value: 71 },
    { month: "Mar-2018", value: 96 },
    { month: "Apr-2018", value: 80 },
    { month: "May-2018", value: 115 },
    { month: "Jun-2018", value: 146 },
    { month: "Jul-2018", value: 121 },
    { month: "Aug-2018", value: 160 },
    { month: "Sep-2018", value: 168 },
    { month: "Oct-2018", value: 185 },
    { month: "Nov-2018", value: 203 }
  ]
};
