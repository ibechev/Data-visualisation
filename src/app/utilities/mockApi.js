import { mockAlerts, mockCustomerInfo } from "../utilities/mockValues";

const randomTimeRange = (min, max) => {
  return (Math.floor(Math.random() * (max - min + 1)) + min) * 1000;
};
const simulateLatency = ms => new Promise(resolve => setTimeout(resolve, ms));

export const mockFetchAlerts = async () => {
  await simulateLatency(randomTimeRange(0, 3));
  return mockAlerts.map(alert => ({ ...alert }));
};

export const mockFetchClientsInfo = async () => {
  const { max, data } = mockCustomerInfo;
  await simulateLatency(randomTimeRange(0, 3));
  return {
    max,
    data: data.map(each => ({ ...each }))
  };
};
