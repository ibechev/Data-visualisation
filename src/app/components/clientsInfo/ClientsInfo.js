import React, { Component, Fragment } from "react";
import { mockFetchClientsInfo } from "../../utilities/mockApi";

import SemiDonutChart from "../semiDonutChart/SemiDonutChart";
import BarChart_12 from "../barChart_12/BarChart_12";
import GrowthBox from "../growthBox/GrowthBox";
import Button from "../button/Button";
import LoadingSpinner from "../loadingSpinner/LoadingSpinner";
import Error from "../error/Error";

class ClientsInfo extends Component {
  constructor(props) {
    super(props);

    this.fetchData = this.fetchData.bind(this);
    this.handleButtonClick = this.handleButtonClick.bind(this);

    this.state = {
      loading: true,
      error: false,
      fetchedData: []
    };
  }

  componentDidMount() {
    this.fetchData();
  }

  handleButtonClick() {
    //eslint-disable-next-line
    console.log("Button was clicked.");
  }

  async fetchData() {
    await mockFetchClientsInfo()
      .then(({ max, data }) => {
        this.setState(prevState => ({
          ...prevState,
          loading: false,
          data,
          max
        }));
      })
      .catch(err => {
        this.setState(prevState => ({ ...prevState, error: true }));
        if (DEVELOPMENT) throw new Error(err);
      });
  }

  render() {
    const { max, data, loading, error } = this.state;

    return (
      <section className="clients-info">
        <header className="clients-info-header">
          <Button noBody onClick={this.handleButtonClick}>
            Clients
            <i className="fas fa-caret-right" />
          </Button>
          <Button noBody onClick={this.handleButtonClick}>
            <i className="fas fa-cog" />
          </Button>
        </header>

        <div className="clients-info-charts">
          {error ? (
            <Error errorText="Oops, there was a problem" />
          ) : loading ? (
            <LoadingSpinner title="information" />
          ) : (
            <Fragment>
              <SemiDonutChart
                name="active clients"
                score={data[data.length - 1].value}
                max={max}
              />
              <BarChart_12 name="Last 12 months" max={max} data={data} />
              <GrowthBox
                prevValue={data[0].value}
                currentValue={data[data.length - 1].value}
              />
            </Fragment>
          )}
        </div>
      </section>
    );
  }
}

export default ClientsInfo;
