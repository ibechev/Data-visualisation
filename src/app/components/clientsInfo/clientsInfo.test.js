import React from "react";
import { shallow, mount } from "enzyme";
import ClientsInfo from "./ClientsInfo";
import renderer from "react-test-renderer";
import { mockCustomerInfo } from "../../utilities/mockValues";

describe("Expect <ClientsInfo />", () => {
  const defaultState = {
    loading: true,
    error: false,
    fetchedData: {
      data: [...mockCustomerInfo.data],
      max: mockCustomerInfo.max
    }
  };

  const setup = (testState = {}) => {
    const state = {
      ...defaultState,
      ...testState
    };
    const wrapper = shallow(<ClientsInfo />);
    wrapper.setState({ ...state });
    return wrapper;
  };

  it("to render correctly", () => {
    const wrapper = renderer.create(<ClientsInfo />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("to call moethod 'fetchData()' after the component mounts", () => {
    const spyFetchData = jest.spyOn(ClientsInfo.prototype, "fetchData");
    shallow(<ClientsInfo />);
    expect(spyFetchData).toHaveBeenCalledTimes(1);
  });

  it("to set 'state.loading' to 'true' and 'state.error' to 'false' by default", () => {
    const wrapper = shallow(<ClientsInfo />);
    expect(wrapper.state("loading")).toEqual(true);
    expect(wrapper.state("error")).toEqual(false);
  });

  describe("when 'state.error' is set to 'true'", () => {
    const wrapper = setup({ error: true });

    it("to not display <LoadingSpinner />", () => {
      expect(wrapper.find("LoadingSpinner")).toHaveLength(0);
    });

    it("to not display <SemiDonutChart />", () => {
      expect(wrapper.find("SemiDonutChart")).toHaveLength(0);
    });

    it("to not display <BarChart_12 />", () => {
      expect(wrapper.find("BarChart_12")).toHaveLength(0);
    });

    it("to not display <GrowthBox />", () => {
      expect(wrapper.find("GrowthBox")).toHaveLength(0);
    });

    it("to display <Error />", () => {
      expect(wrapper.find("Error")).toHaveLength(1);
    });
  });

  describe("when 'state.error' is 'fasle' and 'state.loading' is 'true'", () => {
    const wrapper = setup();

    it("to display <LoadingSpinner />", () => {
      expect(wrapper.find("LoadingSpinner")).toHaveLength(1);
    });

    it("to not display <SemiDonutChart />", () => {
      expect(wrapper.find("SemiDonutChart")).toHaveLength(0);
    });

    it("to not display <BarChart_12 />", () => {
      expect(wrapper.find("BarChart_12")).toHaveLength(0);
    });

    it("to not display <GrowthBox />", () => {
      expect(wrapper.find("GrowthBox")).toHaveLength(0);
    });

    it("to not display <Error />", () => {
      expect(wrapper.find("Error")).toHaveLength(0);
    });
  });

  describe("when 'state.error' is 'fasle' and 'state.loading' is 'false'", () => {
    const wrapper = setup({ loading: false });

    it("to not display <LoadingSpinner />", () => {
      expect(wrapper.find("LoadingSpinner")).toHaveLength(0);
    });

    it("to not display <Error />", () => {
      expect(wrapper.find("Error")).toHaveLength(0);
    });

    it("to display <SemiDonutChart />", () => {
      expect(wrapper.find("SemiDonutChart")).toHaveLength(1);
    });

    it("to display <BarChart_12 />", () => {
      expect(wrapper.find("BarChart_12")).toHaveLength(1);
    });

    it("to display <GrowthBox />", () => {
      expect(wrapper.find("GrowthBox")).toHaveLength(1);
    });
  });

  it("to call button handlers on each click of a button", () => {
    const spyBtnHandler = jest.spyOn(
      ClientsInfo.prototype,
      "handleButtonClick"
    );
    const wrapper = mount(<ClientsInfo />);
    wrapper.find("button").forEach(button => button.simulate("click"));
    expect(spyBtnHandler).toHaveBeenCalledTimes(2);
  });
});
