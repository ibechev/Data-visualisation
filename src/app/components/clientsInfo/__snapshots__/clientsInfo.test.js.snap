// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <ClientsInfo /> to render correctly 1`] = `
<section
  className="clients-info"
>
  <header
    className="clients-info-header"
  >
    <button
      className="button no-body"
      onClick={[Function]}
      type="button"
    >
      Clients
      <i
        className="fas fa-caret-right"
      />
    </button>
    <button
      className="button no-body"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-cog"
      />
    </button>
  </header>
  <div
    className="clients-info-charts"
  >
    <div
      className="loading-spinner"
    >
      <i
        className="fas fa-spinner fa-spin"
      />
      <p
        className="loading-title"
      >
        Loading 
        information
         ...
      </p>
    </div>
  </div>
</section>
`;
